core/o11y/src/log_counter.rs

```diff
@@ -0,0 +1,97 @@
+    try_create_int_counter_vec(
+        "near_log_msg_with_loc_total",
+        "Number of messages logged at various log levels wth target and location",
+        &["level", "target", "location"],
+    )
+    .unwrap()
+});
+
+pub(crate) type LogCountingLayer<Inner> = Layered<LogCounter, Inner>;
+
+/// A tracing Layer that updates prometheus metrics based on the metadata of log events.
+pub(crate) struct LogCounter {
+    // Initializes individual counters as a performance optimization.
+    error_metric: IntCounter,
+    warn_metric: IntCounter,
+    info_metric: IntCounter,
+    debug_metric: IntCounter,
+    trace_metric: IntCounter,
+}
+
+impl LogCounter {
+    /// Increments a counter for every log message.
+    fn count_log(&self, level: &tracing::Level) {
+        match level {
+            &tracing::Level::ERROR => self.error_metric.inc(),
+            &tracing::Level::WARN => self.warn_metric.inc(),
+            &tracing::Level::INFO => self.info_metric.inc(),
+            &tracing::Level::DEBUG => self.debug_metric.inc(),
+            &tracing::Level::TRACE => self.trace_metric.inc(),
+        };
+    }
+
+    /// Increments a counter with target and LoC for high severity messages.
+    fn count_log_with_loc(&self, level: &tracing::Level, target: &str, location: &str) {
+        match level {
+            &tracing::Level::ERROR | &tracing::Level::WARN => LOG_WITH_LOCATION_COUNTER
+                .with_label_values(&[&level.as_str(), target, location])
+                .inc(),
+            // Retaining LoC for low-severity messages can lead to excessive memory usage.
+            // Therefore, only record LoC for high severity log messages.
+            _ => {}
+        };
+    }
+
+    // Event names are usually "event path/to/file.rs:123". Strip the prefix.
+    fn parse_loc<'a>(&'a self, name: &'a str) -> &'a str {
+        if let Some((_, loc)) = name.split_once("event ") {
+            loc
+        } else {
+            name
+        }
+    }
+}
+
+impl Default for LogCounter {
+    fn default() -> Self {
+        Self {
+            error_metric: LOG_COUNTER.with_label_values(&[&OpenTelemetryLevel::ERROR.as_ref()]),
+            warn_metric: LOG_COUNTER.with_label_values(&[&OpenTelemetryLevel::WARN.as_ref()]),
+            info_metric: LOG_COUNTER.with_label_values(&[&OpenTelemetryLevel::INFO.as_ref()]),
+            debug_metric: LOG_COUNTER.with_label_values(&[&OpenTelemetryLevel::DEBUG.as_ref()]),
+            trace_metric: LOG_COUNTER.with_label_values(&[&OpenTelemetryLevel::TRACE.as_ref()]),
+        }
+    }
+}
+
+impl<S> Layer<S> for LogCounter
+where
+    S: tracing::Subscriber + for<'span> LookupSpan<'span> + Send + Sync,
+{
+    fn on_event(&self, event: &tracing::Event, _ctx: Context<S>) {
+        let level = event.metadata().level();
+        self.count_log(level);
+
+        let target = event.metadata().target();
+        let loc = self.parse_loc(event.metadata().name());
```

Comment:
Um, any reason this is not using the dedicated methods to extract the file and line info? 

* https://docs.rs/tracing/latest/tracing/struct.Metadata.html#method.file
* https://docs.rs/tracing/latest/tracing/struct.Metadata.html#method.line

Getting there is pretty convoluted, but ultimately the `tracing_core::metadata!` macro is getting called for `event!`s and `span!`s and [it populates these fields unconditionally](https://github.com/tokio-rs/tracing/blob/b8c45ccd067a6aebdd9257c5edda6bc90bcf9028/tracing-core/src/lib.rs#L262-L271).